veps/sig-storage/incremental-backup.md

```diff
@@ -0,0 +1,330 @@
+
+### Offline backups
+
+Another challenge that needs to be addressed is taking offline backups. Currently, there is no option to create offline backups in Libvirt and QEMU. Backups require QEMU, as it is necessary to interpret the QCOW2 metadata/bitmap.
+The proposed solution is to start the VM in a paused state to take the backup, and then either unpause or shut it down once the process is complete.
+
+### State interruptions during backup
+
+According to the previous solution, if the VM was offline when the backup was requested, we will start the VM in a paused state. When the VM is in a paused state, there is no issue with starting it during the backup process, as Libvirt and QEMU guarantee that any VM operations, such as overwriting blocks in the data file, will not affect the backup's consistency. This consistency assurance could not have been guaranteed if an alternative, non-built-in solution had been considered.
+
+In the case of an online backup if the user tries to stop the VM there will be a built-in [solution](https://issues.redhat.com/browse/RHEL-8067) that Libvirt is currently working on. Until then, we may need to handle it on the Kubevirt side by preventing changes to the VM state during the backup.
+
+## API Examples
+
+### Enable/disable QEMU backup
+
+A new `QCOW2Overlay` field will be added to the VirtualMachine and VirtualMachineInstance CRs. In the VirtualMachine CR, it will be added under: `Spec.Template.Spec.Domain.Devices.QCOW2Overlay`. Similarly, in VirtualMachineInstance CR, it will be added under: `Spec.Domain.Devices.QCOW2Overlay`.
+The field will be of type \*bool and will determine whether the QCOW2 overlay should be applied. As with any other field in the VMI template, this change will only take effect when the VM is restarted. Upon restart, the VMI spec will be created with the updated field value.
+
+When the field is present and set to `true`, a VM state PVC will be created if doesn't already exist. A QCOW2 image will then be created using the existing raw disk image as its data-file. Before applying the domain XML and starting it in virt-launcher manager, the XML will be modified to use the newly created QCOW2 image as the disk.
+
+The XML will modified from:
+```xml
+    <disk type='file' device='disk' model='virtio-non-transitional'>
+      <driver name='qemu' type='raw' cache='none' error_policy='stop' discard='unmap'/>
+      <source file='/var/run/kubevirt-private/vmi-disks/datavolumedisk/disk.img' index='2'/>
+      ...
+    </disk>
+```
+
+To:
+```xml
+    <disk type='file' device='disk' model='virtio-non-transitional'>
+      <driver name='qemu' type='qcow2' cache='none' error_policy='stop' discard='unmap'/>
+      <source file='/run/kubevirt-private/libvirt/qemu/swtpm/qcow2overlay.qcow2' index='2'>
```

Comment:
Better to use the volume name here because a VM could have multiple disks and therefore multiple overlays.